kroxylicious/src/main/java/io/kroxylicious/proxy/codec/CorrelationManager.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.kroxylicious.proxy.codec;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Manages correlation ids for a single connection (across the proxy) between a single client
+ * and a single broker.
+ */
+public class CorrelationManager {
+
+    // TODO use a specialized map
+    /* test */ final Map<Integer, Correlation> brokerRequests = new HashMap<>();
+
+    private int upstreamIds;
+
+    public CorrelationManager() {
+        this(0);
+    }
+
+    public CorrelationManager(int initialCorrelationId) {
+        upstreamIds = initialCorrelationId;
+    }
+
+    /**
+     * Allocate and return a correlation id for an outgoing request to the broker.
+     * @param apiKey The
+     * @param apiVersion
+     * @param downstreamCorrelationId
+     * @param decodeResponse
+     * @return
```

Comment:
> This is how I've been using the words 

Thanks for clarifying. Because I wasn't quite sure and intuitively would have used them exactly the other way around :) I've captured your definition in https://github.com/kroxylicious/design/blob/main/concepts.asciidoc over in the the _design_ repo. As you say, we can grow that into more complete docs later on, for now let's add any definitions and other concepts to it.